Add multi-sig delegation to design document

kantp opened the pull request:
This is the first part of an overhaul of the design document, to catch up with what we have already added to the formal spec and implementation: multi-sig delegation.

I have added script credentials for payments and delegations, and re-written the first part off the document, up to and including the chapter on addresses. It is much clearer now, with payment addresses, stake addresses, and corresponding credentials.

I still need to go over the rest of the document to fix some wording that now might be inconsistent with the new description of addresses.

Commit:
Multi-sig delegation I

Updated chapter on addresses.

Commit:
Remove section on script addresses.

We now do have a (limited) form of script addresses.

Commit:
Adapt rest of the address structure section

With adding multi-sig addresses, we'll need to update how HD wallets work. Made
a todo item for that. There is a way to get HD wallets to work with the
script-based multi-sig addresses, using a slight variation of BIP-45. I've
discussed that with @dcoutts, but still need to write down the particulars.

Commit:
Stake address registration certificates

Commit:
Rewrite section on general address structure.

I think this makes the distinction between single key addresses, general script
addresses, and multi signature addresses, as well as between addresses, value
addresses, and stake addresses, clearer.

Commit:
Update stake pool registration certificates for multi-sig

Commit:
Update delegation certificates for multi-sig

Commit:
Update operational certificates.

In particular, avoid calling the keys stake keys, since they are not.

Commit:
Update Stake pool registration.

Commit:
Update delegation relation.

Commit:
Start removing the term staking key

Commit:
WIP: Work on the diagrams

Commit:
Grammar

dnadales committed:
Update Figures 1 and 2 [skip ci]

dcoutts committed:
Add Matthias to list of contributors

dcoutts committed:
Extend the requirements to cover multi-sig addresses

As well as multi-sig delegation. Yes it's true that the main focus of
Shelley is delegation etc, but we are also now introducing multi-sig
normal addresses, so we should state both.

dcoutts committed:
Update the Overview of Delegation section for scripts

Clarify address vs value address. Add sections on key addresses vs
scripts. All at a high level, not too much detail.

I removed the detail about the payment key pair, since we don't need the
maths in the overview section, and if we kepy it, it would have to be
extended to cover the script case too, which is then far too much detail
for an overview.

dcoutts committed:
Add TODO note on structure of addresses section

And document our agreed terminology for addresses.

Commit:
Spelling correction

Commit:
Rework 3.1

* Change nomenclature to have value and stake addresses form addresses
* Shorten the section a bit
* Removed diagrams -- they would need to be changed significantly, and the whole
design is simpler now, so I'm not sure whether they would still be that useful.
We might add them again later, updated to the new terminology.

Commit:
Rewrite Chapter on Addresses.

It's more systematic now, with payment addresses and stake addresses. Also added
a diagram to show how addresses and credentials are related.

Still TODO: make the wording in the rest of the document consistent.

Commit:
Fix references.

Commit:
Fill in how we satisfy requirement of multi-sig delegation.

dcoutts commented on shelley/design-spec/addresses.dot:
```diff
@@ -0,0 +1,52 @@
+  stake_address [label = "Stake Address"];
+
+  payment_credential [label = "Payment Credential"];
+  stake_credential [label = "Stake Credential"];
```
Wasn't our terminology for this the "stake address reference"? It's where a payment address is referring to a stake address (by value, by pointer/ref, or none).

Note otherwise that we have two things called a stake credential.

JaredCorduan replied:
In the formal spec and exec spec, we are saying that and address has a payment credential and a stake reference. Credentials are either a keyhash or a scripthash. The stake reference is either a credential, a pointer, or null.

kantp replied:
Right! I'll make that change.

kevinhammond replied:
This needs to be reflected into the tutorials, CLI and code base (these are what you are calling base, pointer or enterprise addresses, right?)

kantp replied:
Yes, we have

- base address = stake address reference by value
- pointer address = stake address reference by pointer
- enterprise address = null stake address reference.

Commit:
Add titlesec to default.nix

dcoutts commented on shelley/design-spec/delegation_design_spec.tex:
```diff
@@ -545,276 +573,197 @@ \section{Design of Delegation}
+will distinguish between \emph{value addresses} that determine how funds can be
+spent, and \emph{stake addresses} that define if and how those funds take part
+in the PoS protocol. Actual coins will belong to an (overall) \emph{address},
+consisting of both a value address and a stake address.
```
I think this is not how we currently describe it to ourselves and not the clearest way to think about it.

As in the dot diagram, there are only two kinds of addresses stake addresses and normal addresses for which we should pick a standard term. I've been using payment address recently, but value address is ok. We originally just said "address" but that was before we introduced stake addresses, and then "address" vs "stake address" feels unclear, whereas "payment address" vs "stake address" is much clearer.

So I think we should describe it thus:

> We distinguish between \emph{payment addresses} that determine how funds can be spent, and \emph{stake addresses} that define if and how the stake rights of those funds take part in the PoS protocol. Coins belong to payment addresses. Each payment address (optionally) referrs to a stake address. This delegates the stake rights of any funds held at the payment address to the corresponding stake address. The stake address delegates to a stake pool that participates directly in the PoS protocol. Thus overall there are two steps to the delegation of stake rights: a payment address refers to a stake address; and the stake address delegates to a stake pool.

kantp replied:
Oh right! I should have gone over the introduction again after finishing the chapter on address structure.

dcoutts commented on shelley/design-spec/delegation_design_spec.tex:
```diff
@@ -545,276 +573,197 @@ \section{Design of Delegation}
+  pool. This uniform architecture, not distinguishing between those stakeholders
+  that are using their stake directly and those that are delegating, reduces the
+  overall complexity of the system significantly.
+\end{description}
```
Perhaps worth noting here something like:

> Delegation is optional, but it is the only way to exercise the stake rights to take part in the PoS protocol and to earn rewards.

kevinhammond replied:
Say "Proof-of-Stake" rather than PoS (unfortunate clash of acronym)

Commit:
rename 'stake credential' to 'stake address reference' in payment address.

Also, rename 'by reference' to 'by pointer', to use different words, and make it
consistent with pointer addresses.

Commit:
Clear up payment and stake addresses in introduction.

Commit:
Add note in introduction that delegation is optional.

kevinhammond commented on shelley/design-spec/delegation_design_spec.tex:
```diff
@@ -545,276 +573,205 @@ \section{Design of Delegation}
+counted towards the active stake (\cref{overall-stake-distribution}), and so
+will not slow down chain growth. Users who do not wish to contribute to the PoS
+protocol should use this option. Such addresses are also called \emph{enterprise
+  addresses}.
```
Maybe move away from enterprise addresses now?

kantp replied:
Given that we have used the term for a long time, and it has permeated, I think the document should at least explain the term at this point. But we can try to avoid using it too much across the document and elsewhere.

kevinhammond commented on shelley/design-spec/delegation_design_spec.tex:
```diff
@@ -545,276 +573,205 @@ \section{Design of Delegation}
-The staking rights associated with funds held in this address may be
-exercised by the owner of \(sks\).
+Allowing the stake address reference to be included in a payment address via pointer
+allows for shorter addresses, which is a requirement (\cref{short-addresses}).
```
Short explanation of why a requirement.

kevinhammond commented on shelley/design-spec/delegation_design_spec.tex:
```diff
@@ -545,276 +573,205 @@ \section{Design of Delegation}
+stake address it points to is deregistered, registering the same stake address
+again does not ``restore'' the stake in the pointer address; the tokens have to
+be moved to another address in order to use the stake. This minor limitation
+allows the system to not remember unregistered stake addresses.
```
Not completely clear.  Maybe a diagram would help?

kevinhammond commented on shelley/design-spec/delegation_design_spec.tex:
```diff
@@ -852,106 +804,83 @@ \subsubsection{Reward Account Address}
+\cref{certificate-replay-prevention}. Stake associated with funds in a reward
+account will contribute to the stake of the stake address, so there is no
+incentive to frequently withdraw rewards.
 
```
There may also be a fee for withdrawal, further incentivising redelegation.

kevinhammond commented on shelley/design-spec/delegation_design_spec.tex:
```diff
@@ -852,106 +804,83 @@ \subsubsection{Reward Account Address}
+The terms \emph{reward address}, \emph{reward account}, and \emph{reward account
+  address} will be used interchangeably throughout this document. \todo{Use
+  reward account everywhere instead.}
 
```
Yes, global search and replace, please!!

kevinhammond commented on shelley/design-spec/delegation_design_spec.tex:
```diff
@@ -852,106 +804,83 @@ \subsubsection{Reward Account Address}
+In particular, this kind of wallet scheme allows implementations that can
 do wallet restoration from seed in time that is better than linear in
 the total number of addresses in the blockchain. For details, see
 \cref{wallet-recovery-process}.
```
What does better than linear actually mean here?  Strictly sub-linear is often used as a term, but you might mean asymptotically linear (but often better)?  If it's log or sqrt, say so...

kevinhammond commented on shelley/design-spec/delegation_design_spec.tex:
```diff
@@ -852,106 +804,83 @@ \subsubsection{Reward Account Address}
+stake credential, and check whether it is set according
 to the current delegation preference of the wallet. If there is a
 discrepancy, it will alert the user, and ask them whether they want to
 re-delegate according to their current delegation preferences.
```
This appears to be describing essential requirements for any wallet, rather than what Daedalus does?

If so use the word "must", and highlight somewhere (e.g. as a subsection or concordance) what are the minimal functional requirements for a valid wallet.

kevinhammond commented on shelley/design-spec/delegation_design_spec.tex:
```diff
@@ -1168,44 +1096,44 @@ \subsubsection{Stake Pool Registration Certificates}
+  by using a stake address that is controlled by that charity as the reward
+  address. They can then advertise to other stakeholders that they are doing so
+  (although evidence that the stake address does indeed belong to the charity
+  has to be provided out of band).
```
Does this matter from the perspective of delegation design?  Should it be a description of a general mechanism rather than specific to charities?

kevinhammond commented on shelley/design-spec/delegation_design_spec.tex:
```diff
@@ -1168,44 +1096,44 @@ \subsubsection{Stake Pool Registration Certificates}
-  certificate is required to do so.
+  that adding a stake address to the set of owner stake addresses in itself does
+  not actually delegate the stake controlled by that address to the pool -- an
+  additional delegation certificate is required to do so.
```
Say that the certificate must be registered on chain?

kevinhammond commented on shelley/design-spec/delegation_design_spec.tex:
```diff
@@ -1248,15 +1179,14 @@ \subsubsection{Stake Pool Registration Certificates}
-It must be signed by the staking key \(sks_\text{pool}\) of the pool
-(particularly, it need not be signed by any of the pool owners).
+It must be signed by the pool key \(sk_\text{pool}\) (particularly, it need not
+be signed by any of the pool owners).
```
Is pool key defined?  This is one of the points of potential confusion that the pool operators have spotted

kantp replied:
I think it is, but I'll make it more explicit to be sure.

kevinhammond commented on shelley/design-spec/delegation_design_spec.tex:
```diff
@@ -1295,8 +1225,8 @@ \subsubsection{Single Operator, Possibly Multiple Owners}
-associated with the reward staking key of the pool, and it will be the
-responsibility of the person holding that key (usually the pool operator) to
+associated with the reward address of the pool, and it will be the
+responsibility of the person(s) owning that address (usually the pool operator) to
```
Why does the owner of an address need to be a person?

kevinhammond commented on shelley/design-spec/delegation_design_spec.tex:
```diff
@@ -1363,21 +1293,21 @@ \subsubsection{Operational Key Certificates}
 any information in the 10 corresponding block bodies.
 
-Without this information, the validity of the staking keys that signed
+Without this information, the validity of the hot key that signed
```
Make sure to define hot/cold keys somewhere

kevinhammond commented on shelley/design-spec/delegation_design_spec.tex:
```diff
@@ -1363,21 +1293,21 @@ \subsubsection{Operational Key Certificates}
 \emph{operational key certificates}. These certificates are provided
 in the block header itself, as part of the witness, thus solving the
-problem of determining the validity of staking keys. In other words,
+problem of determining the validity of hot keys. In other words,
```
Here, we had stake addresses, now we have hot keys.  Is that a fundamental change or just terminology (i.e. are these hot stake keys, hot pool keys, or what?).   Also, I would think that a valid hot key is one that has been created from a cold key and is within its period of validity.  Is that what is meant, or something else?

kantp replied:
Re. 1: basically, we've introduced more types to the delegation system as we progressed with the design. At first, we had stake keys for different things, and delegations between them, and special rules restricting chain delegations. Now, we have stake addresses, which delegate to pool keys, which issue certificates for hot keys (and no special rules restricting chain delegations, since there is only one possible chain respecting the types).

Re. 2: what I really meant is _precedence_ of hot keys, so I'll change that :)

kevinhammond commented on shelley/design-spec/references.bib:
```diff
@@ -131,6 +131,15 @@ @misc{bip32
+  url    = {https://github.com/bitcoin/bips/blob/master/bip-0045.mediawiki},
+  note   = {BIP-45},
+}
+
```
Ryan X. Charles?

kevinhammond commented on shelley/design-spec/references.bib:
```diff
@@ -176,3 +185,9 @@ @article{DBLP:journals/tcs/Escardo96
+    author = {Jared Corduan and Matthias G\"{u}demann},
+    title = {A Formal Specification of a Multi-Signature Scheme using Scripts},
+    year = {2019}
+}
```
Add a proper reference here. At least a URL or SL-Dx reference.

kevinhammond reviewed (approved):
I have looked through this and made some comments.  I think this should go ahead but please consider the comments.   Note that PoS is a very unfortunate abbreviation. It should be deprecated.  We should also have a glossary somewhere.
